### PR TITLE
[FEAT] F.scaled_dot_product_attention backend for GQA/MLA

### PR DESCRIPTION
**Pull request by mvanhorn**
## Summary

Adds a `use_sdpa: bool = True` flag to `MythosConfig`. When set (the default), `GQAttention.forward` and `MLAttention.forward` dispatch to `torch.nn.functional.scaled_dot_product_attention` instead of computing softmax by hand. On CUDA this automatically picks up Flash Attention and memory-efficient attention kernels, no extra config.

KV-cache handling and RoPE application are unchanged. The swap happens strictly after K/V reconstruction in MLA's forward and after `repeat_interleave` in GQA's, so the tensors fed to SDPA are exactly what the manual softmax path receives today.

`use_sdpa=False` keeps the hand-written softmax available for debugging or study, since this is an educational-research repo.

## Why this matters

The recurrent block runs attention `T` times per forward pass (prelude once + recurrent loop N times + coda once), so attention cost compounds quickly at inference. SDPA is the PyTorch team's recommended attention entry point since 2.0, and `torch>=2.1.0` is already this project's minimum. The handwritten softmax blocks were ~5 lines each but left Flash Attention on the table.

### `open_mythos/main.py` (+35, -12)

- `MythosConfig`: added `use_sdpa: bool = True` (documented in the docstring).
- `GQAttention.__init__`: stores `self.use_sdpa = cfg.use_sdpa`.
- `GQAttention.forward`: branches on `self.use_sdpa`. SDPA path passes the additive causal mask via `attn_mask=mask` (identical math to adding mask before softmax). Fallback path is byte-for-byte the original code.
- `MLAttention.__init__` / `forward`: same change. Note MLA's `q_head_dim = qk_nope_head_dim + qk_rope_head_dim` and `v_head_dim` differ - SDPA handles heterogeneous head dims fine.

### `test_main.py` (+46, -0)

Added `test_sdpa_equivalence` to both `TestGQAttention` and `TestMLAttention`. Each test:
1. Builds two attention modules with identical state_dict (one `use_sdpa=False`, one `use_sdpa=True`).
2. Runs both on the same input tensor under `torch.no_grad()`.
3. Asserts `torch.allclose(manual, sdpa, atol=1e-5)`.

Measured diff in a CPU dev run: `1.79e-07` for GQA, `~3e-07` for MLA - well within float32 rounding.

## Testing

CPU, Python 3.14, torch 2.11.0:

```
$ pytest test_main.py -k test_sdpa_equivalence -v
test_main.py::TestGQAttention::test_sdpa_equivalence PASSED    [ 50%]
test_main.py::TestMLAttention::test_sdpa_equivalence PASSED    [100%]
2 passed in 0.34s

$ pytest test_main.py
68 passed, 1 failed (pre-existing: test_spectral_radius_stable_after_large_grad_step flakes at the clamp boundary on upstream main too - unrelated to this PR)

$ python example.py
[MLA] Parameters: 1,538,626
[MLA] Logits shape: torch.Size([2, 16, 1000])
[MLA] Generated shape: torch.Size([2, 24])
[MLA] Spectral radius p(A) max: 0.3679 (must be < 1)
```

End-to-end speedup comparison (CPU, dim=512, seq=128, 4 loops):

| Backend | ms/forward |
|---|---|
| manual softmax | 15.55 |
| SDPA | 14.37 |

1.08x on CPU is modest - the real win is on CUDA where SDPA dispatches to `FLASH_ATTENTION` / `EFFICIENT_ATTENTION` kernels automatically. I don't have a GPU locally so I can't benchmark that here, but per `torch.backends.cuda.sdp_kernel` docs the fused kernels kick in for causal masked attention with no intervention.

## Demo

sdpa-equivalence-and-speedup

Four frames: new tests pass, `use_sdpa` flag is wired in MythosConfig, CPU benchmark (1.08x, max abs diff 2.15e-6 on a full forward), example.py still prints the advertised output.

## Alternatives Considered

1. Swap unconditionally and delete the manual path. Rejected - this is a research-code repo where people likely step through attention to study it. The flag is cheap to keep.
2. Add a hard `flash-attn` dependency. Rejected - doesn't build on macOS/Windows without CUDA. SDPA handles dispatch automatically with no extra imports.

No upstream issue - this PR is the proposal.

This contribution was developed with AI assistance (Codex).

### CODE REVIEW

**Comment by kyegomez** [comment]
@mvanhorn hey i'd like to merge this but there are some conflicts that we need to solve first

**Comment by mvanhorn** [comment]
Rebased onto main in db3ac8d. The two conflicts were in `GQAttention.forward` and `MLAttention.forward` where upstream re-added `attn_drop` wrapping `F.softmax` (eae0f04). Kept both changes: the if/else structure stays, and the non-SDPA fallback now uses `attn_drop(F.softmax(...))`. For the SDPA path, I plumbed `dropout_p` via `self.attn_drop.p` during training (0 in eval) so dropout is preserved through the SDPA kernel. Should be directly mergeable now.

**Comment by mvanhorn** [comment]
Rebased onto main in 2f2ef5b. Upstream added the `_HAS_FLASH_ATTN` top-level branch and renamed `self.attn_drop` to `self.dropout_p` in GQAttention, so the conflict resolution places `use_sdpa` inside the `else` (fallback) path and uses `self.dropout_p`. Also updated the two `test_sdpa_equivalence` tests to pass `dtype=torch.float32` to `_causal_mask` (new required arg). Both tests pass (GQA and MLA equivalence within 3e-07).

**Comment by Qodo-Free-For-OSS** [comment]
Hi, In `GQAttention.forward` and `MLAttention.forward`, the SDPA path sets `is_causal=True` whenever `mask is None` and `T>1`, changing the semantics of `mask=None` from unmasked/bidirectional attention to causal attention. This can produce incorrect outputs for any callers that intentionally omit the mask (and it also diverges from the manual path, which applies masking only when `mask is not None`).

**Severity:** action required | **Category:** correctness

**How to fix:** Align SDPA mask semantics

**Agent prompt to fix** - you can give this to your LLM of choice:

> ### Issue description
> `GQAttention.forward` and `MLAttention.forward` set `is_causal=True` when `mask is None` and `T>1`. Historically in this repo, `mask=None` means *no mask is applied* (bidirectional attention). The SDPA path therefore changes behavior vs the manual path.
> 
> ### Issue Context
> - Manual path only adds mask when `mask is not None`.
> - `OpenMythos.forward` always provides a causal mask for `T>1`, so the issue is mostly visible for external callers/tests calling attention modules directly.
> 
> ### Fix Focus Areas
> - open_mythos/main.py[272-281]
> - open_mythos/main.py[429-438]
> 
> ### What to change
> - Remove the implicit causality when `mask is None` (i.e., set `is_causal=False` when `mask is None`), **or** update the manual path to match the new semantics (but that would be a broader behavior change).
> - If you still want SDPA to use causal masking for the model’s normal path, rely on the already-supplied additive causal mask from `OpenMythos._causal_mask` (pass via `attn_mask`) rather than inferring from `mask is None`.

---
*Spotted by [Qodo](https://github.com/marketplace/qodo-merge-pro-for-open-source) code review - free for open-source projects.*

**Comment by mvanhorn** [comment]
Good catch. Pushed 4651e97 setting `is_causal=False` on both SDPA paths so `mask=None` matches the manual path's bidirectional semantics. Callers that want causal attention must supply an explicit mask via `OpenMythos._causal_mask`, which the model's main forward already does for T>1, so the training/inference hot paths are unchanged.

Added `test_sdpa_no_mask_is_bidirectional` regression tests for both GQAttention and MLAttention. All 4 SDPA tests pass on torch 2.x:

```
tests/test_main.py::TestGQAttention::test_sdpa_equivalence PASSED
tests/test_main.py::TestGQAttention::test_sdpa_no_mask_is_bidirectional PASSED
tests/test_main.py::TestMLAttention::test_sdpa_equivalence PASSED
tests/test_main.py::TestMLAttention::test_sdpa_no_mask_is_bidirectional PASSED
```

(Note: 7 unrelated `test_output_shape` / `test_with_causal_mask` tests fail on main with a RoPE freqs sizing mismatch in `apply_rope`. Those exist on the base branch too, unrelated to this PR.)
